### ISSUE TITLE
D3D8 to DXVK already has a working alternative. Do you plan to surpass it?

### ISSUE THREAD

**Issue by LethalManBoob** [AlpyneDreams/d8vk#114]
https://github.com/crosire/d3d8to9
d3d8to9, albeit made for a different purpose, can function very well with dxvk. Simply dropping it in makes the game pass d3d8 to d3d9 and then to dxvk flawlessly in my experience. 
Do you plan to draw influence from it or collab in any way?

**Comment by WinterSnowfall** [comment]
I think your question is already mostly addressed in the latest release notes ([see "Features" here](https://github.com/AlpyneDreams/d8vk/releases/tag/d8vk-v0.10)), however I'll dare to reply as well, as a long time user of d3d8to9 with dxvk. Admittedly d3d8to9 generally works well with dxvk - but the way it handles some things is particularly... "interesting" to say the least.

d8vk (as the name suggests) is built to interface with dxvk's backend, so in that sense it has already surpassed d3d8to9 in terms of compatibility with dxvk from a design standpoint. Performance should already be better with d8vk, not that you generally need to worry a lot about performance with most d3d8 games. Of course there are still plenty of quirks and corner cases yet to handle with regards to the d3d8 implementation, but I believe it's coming along nicely.

P.S.: Mind you, WineD3D's d3d8 to OpenGL conversion is also a quite functional alternative. So is using dgVoodoo 2 with dxvk :).

**Comment by AlpyneDreams** [comment]
> https://github.com/crosire/d3d8to9 d3d8to9, albeit made for a different purpose, can function very well with dxvk. Simply dropping it in makes the game pass d3d8 to d3d9 and then to dxvk flawlessly in my experience.

d3d8to9 is a great project, but its goals differ from d8vk's. d3d8to9 aims to translate API calls directly to d3d9, without the consideration for compatibility or performance that is fundamental to dxvk. They admit this difference in their own README:

>Note that d3d8to9 does exact translation from Direct3D 8 to Direct3D 9 only. But on recent versions of Windows, depending on system settings, GPU drivers and other factors, rendering behavior in Direct3D 9 using d3d8to9 may differ from native Direct3D 8. One common issue is VSync getting enabled with d3d8to9, even though it appears inactive in native Direct3D 8, despite the application enabling it. This may give an inaccurate impression of lower performance with d3d8to9, while it actually follows what the application requests.

d8vk aims to implement native Direct3D 8. Some of the advantages of d8vk include:
- First class support for Linux and Wine
- Better performance in many cases
- Individual configurations to support specific games
- [Support for vendor hacks and driver quirks](https://github.com/AlpyneDreams/d8vk/commit/e3009f68117e22ba4e3f4b838e819597d5d6fdfa)
  - This means games like Splinter Cell will look better with d8vk

Of course, d8vk will not work with native Direct3D 9. While d3d8to9 still currently provides (usually) the best compatibility with d3d8, that comes at a cost of performance and maintainability. A few examples of the limitations of d3d8to9 include:
-  It depends on D3DX which is proprietary nonfree software that cannot be redistributed and is implemented differently by Wine
- It cannot perform many types of texture copies that d3d9 does not support, and therefore has to download those textures to the CPU
- It updates shaders to be d3d9-compatible by decompiling them to source code text and using regex replace to make them compatible, and then re-compiling them.
- It uses lookup tables to map d3d9 objects to d3d8 objects, which can be slower
- Lots of tiny little things like using `0x80000000` instead of `D3DFVF_RESERVED0` to mark non-FVF shaders that can cause issues for some games
- It doesn't call EndScene in Reset, so some games like #111 will break.

> Do you plan to draw influence from it or collab in any way?

All that being said, d3d8to9 has been instrumental in debugging d8vk and their implementation was useful as one of several references in the early days of d8vk. That's why it is included in the acknowledgements of our release notes. It is a fantastic project and I have nothing but respect for crosire and the other contributors.

I also agree with everything @WinterSnowfall said.

**Comment by AlpyneDreams** [comment]
> D3D8 to DXVK already has a working alternative. Do you plan to surpass it?

TL;DR: Yes.

**Comment by elishacloud** [comment]
> A few examples of the limitations of d3d8to9 include:

Not all of those examples are true.  See details below.

> * It depends on D3DX which is proprietary nonfree software that cannot be redistributed and is implemented differently by Wine

D3DX is part of a DirectX runtime and the runtime can be redistributed.  It just that D3DX cannot be redistributed by itself.

> * It cannot perform many types of texture copies that d3d9 does not support, and therefore has to download those textures to the CPU

This is not a limitation of d3d8to9.  This is a limitation of Direct3D9.

> * It updates shaders to be d3d9-compatible by decompiling them to source code text and using regex replace to make them compatible, and then re-compiling them.

True.

> * It uses lookup tables to map d3d9 objects to d3d8 objects, which can be slower

I am curious how else this can be done?  How can an address that the game knows be mapped back to an address for the Direct3D9 associated resource?

> * Lots of tiny little things like using `0x80000000` instead of `D3DFVF_RESERVED0` to mark non-FVF shaders that can cause issues for some games

`0x80000000` is used for determining if the address is one from d3d8to9 or not.  We cannot use `D3DFVF_RESERVED0` because some games use that.  In that case we would not be able to distinguish between an address in d3d8to9 and a handle that the game created using `D3DFVF_RESERVED0`.

> * It doesn't call EndScene in Reset, so some games like https://github.com/AlpyneDreams/d8vk/issues/111 will break.

This is not needed in Direct3D9 and thus not needed in d3d8to9 because when you call Reset() it resets the scene also so BegineScene() will not fail with d3d8to9 like it did in that case.

**Comment by elishacloud** [comment]
I should mention that if you want to go to Vulkan then going directly from DX8 to VK is better than going from DX8 to DX9 and then to VK.  Less wrappers is always better and more efficient.

**Comment by WinterSnowfall** [comment]
@elishacloud To provide some context, a lot of the above is said from the perspective of d3d8 use on Linux and more particularly addressed at people who are using d3d8to9 + dxvk rather than resorting to WineD3D's d3d8 to OpenGL translation, as per the initial post. I was one of them for many years, long before d8vk existed, so I'll take this opportunity to thank you for your work :).

That being said, most of the points above are purely a reflection of differences in implementation and ultimately project goals. Since d8vk has access to the dxvk backend (it esentially _is_ a d3d8/d3d9 implementation on top of Vulkan), it can do a lot more things such as implement texture copies on the GPU, interact directly with VK objects and also apply workarounds for games which are known to be broken due to various reasons either on native or Linux drivers (a simple example of this is limiting the amount of available video memory reported to a game). This is in line with what dxvk has also been doing for years, but with added d3d8 support.

Another thing to generally keep in mind is that d8vk, much like dxvk, doesn't particularly aim to support use on Windows. In most cases it should work fine, but there's no explicit guarantee of compatibility on Windows and native Vulkan drivers. Getting d3d8 games to work with native d3d9 implementations by means of d3d8to9 is most likely a better option when using Windows.

As to the whole EndScene() in Reset() situation, that is now resolved in upstream dxvk and should work fine with d3d8to9 + dxvk, but it did not work properly at the time of writing (the fault was ultimately dxvk's). d3d8to9 + WineD3D will still have issues here however, since WineD3D's d3d9 implementation doesn't handle it properly to this day.

**Comment by elishacloud** [comment]
@WinterSnowfall, Thanks for the context.  Interesting about WineD3D.  I have not done much testing on that.  I mostly focus on Microsoft's implementation of Direct3D.  dxvk and d8vk are probably better for Linux than d3d8to9.  d3d8to9 is primarily used for compatibility in Windows and for using 3rd party tools that don't support older versions of DX, like ReShade and RTX Remix.

BTW: I am also working on a dd7to9 feature that can convert games from DirectDraw/Direct3D7 and older to Direct3D9.  It works similar to how d3d8to9 works.
